### PR TITLE
feat: rewrite config with hook and add rtl config

### PR DESCRIPTION
**Pull request by xrkffgg**
### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [x] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link
close https://github.com/ant-design/ant-design/issues/23728

### 💡 Background and solution
- config-provider rewrite with hook
- Add message/notification rtl config direct

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |         none  |
| 🇨🇳 Chinese |  none         |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️ 

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

**Comment by xrkffgg** [comment]
很奇怪，报一个这个错

image

我把这一行 注释了 就能行了，但我修改的 不应该影响这个啊
https://github.com/ant-design/ant-design/blob/9cd8f793668d8ff71e22b47e3c49eecfa1408802/components/dropdown/dropdown-button.tsx#L11

File: components/config-provider/index.tsx [hunk 1]
```diff
@@ -49,10 +51,21 @@ export interface ConfigProviderProps {
   };
 }
 
-class ConfigProvider extends React.Component<ConfigProviderProps> {
-  getPrefixClsWrapper = (context: ConfigConsumerProps) => {
+const ConfigProvider: React.FC<ConfigProviderProps> = props => {
+  React.useEffect(() => {
+    if (props.direction && typeof props.direction !== undefined) {
```
Comment (yoyo837) [inline components/config-provider/index.tsx:56]: https://lgtm.com/projects/g/ant-design/ant-design/rev/pr-5ec4d7bdfab6edd6b5ade0a73dba5d7b22daf1e0
typeof 返回字符串

Comment (yoyo837) [inline components/config-provider/index.tsx:56]:
```suggestion
    if (props.direction !== undefined) {
```

File: components/config-provider/index.tsx [hunk 2]
```diff
@@ -49,10 +51,21 @@ export interface ConfigProviderProps {
   };
 }
 
-class ConfigProvider extends React.Component<ConfigProviderProps> {
-  getPrefixClsWrapper = (context: ConfigConsumerProps) => {
+const ConfigProvider: React.FC<ConfigProviderProps> = props => {
+  React.useEffect(() => {
+    if (props.direction) {
+      message.config({
+        rtl: props.direction === 'rtl',
```
Comment (yoyo837) [inline components/config-provider/index.tsx:58]: 抽取变量 `const rtl = props.direction === 'rtl'` 复用吧

Comment (xrkffgg) [reply components/config-provider/index.tsx:58]: 暂时就 2 个地方用，还好

**Review by afc163 — APPROVED** [review]

**Comment by afc163** [comment]
@zombieJ 还能赶上车么？

**Comment by afc163** [comment]
@xrkffgg 

这个改动让测试覆盖行数掉了 10 行，对比 master 和 feature 的覆盖：

- master: https://codecov.io/gh/ant-design/ant-design/tree/feature/components/_util
- feature: https://codecov.io/gh/ant-design/ant-design/tree/master/components/_util

比如 `openAnimation.tsx` 这个文件，从 1 行掉到 7 行。

在 `openAnimation.tsx` 的 `active` 方法内加上 `console.log`，在 master 下跑 `npm test -- components/collapse/__tests__/index.test.js` 可以测试：

image

相同的操作在 feature 运行，console.log 没有生效。此时把 `config-provider/index.tsx` 复原回 master 的代码再跑 `npm test -- components/collapse/__tests__/index.test.js`，console.log 就出来了。

**Comment by xrkffgg** [comment]
@hengkx 看下 hooks 改写的有问题吗？

**Comment by hengkx** [comment]
没问题

**Comment by xrkffgg** [comment]
image

我这里 active 竟然都没进去

**Comment by xrkffgg** [comment]
image

**Pull request merged**